src/runtime/webgpu.cpp

```diff
@@ -0,0 +1,1146 @@
+#include "HalideRuntimeWebGPU.h"
+#include "device_buffer_utils.h"
+#include "device_interface.h"
+#include "gpu_context_common.h"
+#include "printer.h"
+#include "scoped_spin_lock.h"
+
+#include "mini_webgpu.h"
+
+namespace Halide {
+namespace Runtime {
+namespace Internal {
+namespace WebGPU {
+
+extern WEAK halide_device_interface_t webgpu_device_interface;
+
+WEAK int create_webgpu_context(void *user_context);
+
+// A WebGPU instance/adapter/device defined in this module with weak linkage.
+WEAK WGPUInstance global_instance = nullptr;
+WEAK WGPUAdapter global_adapter = nullptr;
+WEAK WGPUDevice global_device = nullptr;
+// Lock to synchronize access to the global WebGPU context.
+volatile ScopedSpinLock::AtomicFlag WEAK context_lock = 0;
+
+// Size of the staging buffer used for host<->device copies.
+constexpr int kWebGpuStagingBufferSize = 4 * 1024 * 1024;
+// A staging buffer used for host<->device copies.
+WGPUBuffer WEAK staging_buffer = nullptr;
```

Comment:
Should access to this be protected with a mutex? Seems like you could have a situation where two host threads were in contention to use this for copy_to_host, etc

Reply:
OK I see what you're getting at now. Yes, for the `gpu_multi_context_threaded` tests, we need to create a staging buffer for each context to avoid the contention.